### ISSUE TITLE
Error when building RAID-5 on FEMU drives

### ISSUE THREAD

**Issue by NeverMore2744** [MoatLab/FEMU#138]
I tried to build a RAID-5 system on 7 emulated black-box SSDs with mdadm. After the parity creation finished (RAID-5 starts writing parities upon the RAID creation), the FEMU crashes with the error:

fvm login: [FEMU] FTL-Err: No free lines left in [vSSD6].

Is it related to the overprovisioning of SSDs?
Is it possible to use mdadm to build a RAID system with FEMU emulated drives without errors?

**Comment by huaicheng** [comment]
> Is it related to the overprovisioning of SSDs?

Yes, any misconfigrations in your setup? Make sure the nvme device capacity exposed to the host is less than the raw nand flash capacity.

>Is it possible to use mdadm to build a RAID system with FEMU emulated drives without errors?
Yes, by default it should work fine.

**Comment by NeverMore2744** [comment]
Thanks for your reply. I create the black-box SSD like this: 
> -device femu,devsz_mb=16384,femu_mode=1
It seems that both the capacity exposed to the host and the raw nand flash capacity are 16GiB. If all the capacity is written, it causes the error.
Now I found that I can specify the --size parameter in mdadm so that it does not use all 16GiB for RAID, and mdadm can run normally.

**Comment by huaicheng** [comment]
Thanks — and your own diagnosis was right: the namespace was exactly as large as the raw NAND, so once mdadm's parity pass filled it there was nothing left for garbage collection to reclaim. `--size` was a sound workaround.

FEMU should not have aborted on you, though. A line can only be reclaimed by relocating its valid pages somewhere else, so some of the NAND has to stay unexposed; with `devsz_mb=16384` and the default geometry (which is also 16 GiB) there was no spare at all, and the write path hit `abort()` mid-run.

That is now refused at startup instead, in ceb2b5bf7:

```
$ qemu-system-x86_64 ... -device femu,devsz_mb=16384,femu_mode=1
FEMU bbssd: namespace 1 exposes 16384 MiB of the 16384 MiB this geometry has,
leaving garbage collection no room; expose at most 15552 MiB per namespace
(lower devsz_mb, or set op_pcent)
```

So you get told the number that works rather than a crash after the array has been building for a while. `op_pcent` is the tidier way to express it — `op_pcent=10` sizes the namespace at NAND/1.10 automatically.

Closing; RAID-5 on FEMU drives works fine within a correctly sized namespace.
